client/consensus/manual-seal/src/heartbeat_stream.rs

```diff
@@ -0,0 +1,133 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! HeartbeatStream type and HeartbeatOptions used in instant seal
+use std::{pin::Pin};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
+
+use crate::rpc::{EngineCommand};
+
+/// Constants for generating default heartbeat options
+const DEFAULT_MAX_BLOCKTIME: u64 = 30;
+const DEFAULT_MIN_BLOCKTIME: u64 = 1;
+const DEFAULT_FINALIZE: bool = false;
+
+/// Heartbeat options to manage the behavior of the heartbeat stream
+pub struct HeartbeatOptions {
+	/// The minimum amount of time to pass before generating a new block, in sec.
+	pub min_blocktime: u64,
+	/// The amount of time passed that a new heartbeat block will be generated, in sec.
+	pub max_blocktime: u64,
+	/// Control whether the generated block is finalized
+	pub finalize: bool,
+}
+
+impl Default for HeartbeatOptions {
+	fn default() -> Self {
+		Self {
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			max_blocktime: DEFAULT_MAX_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+/// HeartbeatStream is combining transaction pool notification stream and generate a new block
+///   when a certain time has passed without any transactions.
+pub struct HeartbeatStream<Hash> {
+	/// The transaction pool notification stream
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	/// Future to control when the next block should be generated
+	delay_future: Delay,
+	/// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	/// Heartbeat options
+	opts: HeartbeatOptions,
```

Comment:
```suggestion
	options: HeartbeatOptions,
```

(opinionated)
I'd keep all variable names un-abbreviated except special 1 letter cases.

Reply:
updated